File: quick_sort/ruby/quick_sort.rb
class QuickSort
  def self.sort(list)
    return list if list.size <= 1

    pivot = list[rand(list.size)]

    left = []
    right = []

    list.each do |item|
      if item > pivot
        right.push(item)
      else
        left.push(item)
      end
    end

    sort(left) + sort(right)
  end
end

File: quick_sort/cpp/quick_sort.cpp
#include "quick_sort.h"

void QuickSort::Sort(int numbers[], int left, int right) const
{
  int tmp;
  int i = left, j = right;
  int pivot = numbers[(left + right) / 2];

  while (i <= j) {

    while(numbers[i] < pivot) {
      i++;
    }

    while(numbers[j] > pivot) {
      j--;
    }

    if (i <= j) {
      tmp = numbers[i];
      numbers[i] = numbers[j];
      numbers[j] = tmp;
      i++;
      j--;
    }
  }

  if (left < j) {
    this->Sort(numbers, left, j);
  }

  if (i < right) {
    this->Sort(numbers, i, right);
  }
}
